zwavefibaro/integrationpluginzwavefibaro.cpp

```diff
@@ -0,0 +1,231 @@
+* contact@nymea.io or see our FAQ/Licensing Information on
+* https://nymea.io/license/faq
+*
+* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include "integrationpluginzwavefibaro.h"
+#include "plugininfo.h"
+
+#include <hardware/zwave/zwavehardwareresource.h>
+
+#include <QDebug>
+
+IntegrationPluginZWaveFibaro::IntegrationPluginZWaveFibaro(): ZWaveIntegrationPlugin(ZWaveHardwareResource::HandlerTypeVendor, dcZWaveFibaro())
+{
+}
+
+QString IntegrationPluginZWaveFibaro::name() const
+{
+    return "Fibaro";
+}
+
+bool IntegrationPluginZWaveFibaro::handleNode(ZWaveNode *node)
+{
+    qCWarning(dcZWaveFibaro()) << "Handle node for Fibaro";
+
+    if (node->manufacturerId() != 0x010f) {
+        return false;
+    }
+
+    // FGWPE/F Wall Plug (0x1000)
+    if (node->productId() == 0x1000) {
+        createThing(powerSocketThingClassId, node);
+        return true;
+    }
+
+    // 0x1002 = EU version, 0x2002 = US version
+    if (node->productId() == 0x1002 || node->productId() == 0x2002) {
+        createThing(motionSensorThingClassId, node);
+        return true;
+    }
+
+    return false;
+}
+
+void IntegrationPluginZWaveFibaro::setupThing(ThingSetupInfo *info)
+{
+    Thing *thing = info->thing();
+
+    if (!manageNode(thing)) {
+        qCWarning(dcZWaveFibaro()) << "Failed to claim node during setup.";
+        info->finish(Thing::ThingErrorHardwareNotAvailable);
+        return;
+    }
+
+    ZWaveNode *node = nodeForThing(thing);
+
+    if (thing->thingClassId() == powerSocketThingClassId) {
+
+        thing->setSettingValue(powerSocketSettingsAlwaysOnModeParamTypeId, node->value(ZWaveValue::GenreConfig, ZWaveValue::CommandClassConfiguration, 1, 1, ZWaveValue::TypeList).valueListSelection() == 0);
+        thing->setSettingValue(powerSocketSettingsRestoreModeParamTypeId, node->value(ZWaveValue::GenreConfig, ZWaveValue::CommandClassConfiguration, 1, 16, ZWaveValue::TypeList).valueListSelection() == 1);
+        QVariantList ledModes = thing->thingClass().settingsTypes().findById(powerSocketSettingsLedModeParamTypeId).allowedValues();
+        thing->setSettingValue(powerSocketSettingsLedModeParamTypeId, ledModes.at(node->value(ZWaveValue::GenreConfig, ZWaveValue::CommandClassConfiguration, 1, 61, ZWaveValue::TypeList).valueListSelection()));
+
+        thing->setStateValue(powerSocketPowerStateTypeId, node->value(ZWaveValue::GenreUser, ZWaveValue::CommandClassSwitchBinary, 1, 0, ZWaveValue::TypeBool).value().toBool());
+        thing->setStateValue(powerSocketCurrentPowerStateTypeId, node->value(ZWaveValue::GenreUser, ZWaveValue::CommandClassSensorMultilevel, 1, 4, ZWaveValue::TypeDecimal).value().toDouble());
+        thing->setStateValue(powerSocketTotalEnergyConsumedStateTypeId, node->value(ZWaveValue::GenreUser, ZWaveValue::CommandClassMeter, 1, 0, ZWaveValue::TypeDecimal).value().toDouble());
+
+        info->finish(Thing::ThingErrorNoError);
```

Comment:
Isn't it better to update the states after you finish the setup to prevent the warnings in the core?

Reply:
It doesn't print a warning IIRC, however, setting it before the finish call would cause the state change to *not* appear in the log DB. So yes, I've changed the order. thanks